src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs

```diff
@@ -0,0 +1,355 @@
+        /// <param name="launch">launch (launch context)URL parameter.</param>
+        /// <param name="scope">scope URL parameter.</param>
+        /// <param name="state">state URL parameter.</param>
+        /// <param name="aud">aud (audience) URL parameter.</param>
+        [HttpGet("authorize")]
+        public ActionResult Authorize(
+            [FromQuery(Name = "response_type")] string responseType,
+            [FromQuery(Name = "client_id")] string clientId,
+            [FromQuery(Name = "redirect_uri")] Uri redirectUri,
+            [FromQuery(Name = "launch")] string launch,
+            [FromQuery(Name = "scope")] string scope,
+            [FromQuery(Name = "state")] string state,
+            [FromQuery(Name = "aud")] string aud)
+        {
+            EnsureArg.IsNotNull(responseType, nameof(responseType));
+            EnsureArg.IsNotNull(clientId, nameof(clientId));
+            EnsureArg.IsNotNull(redirectUri, nameof(redirectUri));
+            EnsureArg.IsNotNull(aud, nameof(aud));
+
+            if (string.IsNullOrEmpty(launch))
+            {
+                launch = Base64Encode("{}");
+            }
+
+            JObject newStateObj = JObject.Parse("{}");
+            newStateObj.Add("s", state);
+            newStateObj.Add("l", launch);
+
+            string newState = Base64Encode(newStateObj.ToString(Newtonsoft.Json.Formatting.None));
+
+            Uri callbackUrl = new Uri(
+                Request.Scheme + "://" + Request.Host + "/AadProxy/callback/" +
+                Uri.EscapeDataString(Base64Encode(redirectUri.ToString())));
+
+            string newQueryString = $"response_type={responseType}&redirect_uri={callbackUrl.ToString()}&client_id={clientId}";
+            if (!_isAadV2)
+            {
+                newQueryString += $"&resource={aud}";
+            }
+            else
+            {
+                // Azure AD v2.0 uses fully qualified scopes and does not allow '/' (slash)
+                // We add qualification to scopes and replace '/' -> '$'
+
+                EnsureArg.IsNotNull(scope, nameof(scope));
+                var scopes = scope.Split(' ');
+                string newScopes = string.Empty;
+                string[] wellKnownScopes = { "profile", "openid", "email", "offline_access" };
+
+                foreach (var s in scopes)
+                {
+                    if (wellKnownScopes.Contains(s))
+                    {
+                        newScopes += $"{s} ";
+                    }
+                    else
+                    {
+                        newScopes += $"{aud}/{s.Replace('/', '$')} ";
+                    }
+                }
+
+                newScopes = newScopes.TrimEnd(' ');
+                newQueryString += $"&scope={Uri.EscapeDataString(newScopes)}";
+            }
+
+            newQueryString += $"&state={newState}";
+
+            return RedirectPermanent($"{_aadAuthorizeEndpoint}?{newQueryString}");
```

Comment:
Is this supposed to be a permanent redirect?

Reply:
Switched to `Redirect`